### PR TITLE
Fix index overflow bug in einsum

### PR DESCRIPTION
**Pull request by hzfan**
## Description ##
Fix index overflow bug in einsum as mentioned in #16576

## Comments ##
Thank @reminisce for guidance and review.

### CODE REVIEW

**Comment by hzfan** [comment]
@mxnet-label-bot  add [Numpy]

File: tests/python/unittest/test_numpy_op.py [hunk 1]
```diff
@@ -3377,16 +3377,20 @@ def dbg(name, data):
...
-    dtypes = ['int32', 'float16', 'float32', 'float64']
+    dtypes = ['int32', 'float32', 'float64']
     for hybridize in [False, True]:
         for dtype in dtypes:
             for config in configs:
                 for optimize in [False, True]:
-                    rtol = 1e-0 if dtype == 'float16' else 1e-3
-                    atol = 1e-1 if dtype == 'float16' else 1e-5
+                    rtol = 1e-0 if dtype == 'float16' else 1e-1
+                    atol = 1e-1 if dtype == 'float16' else 1e-1
```
Comment (sxjscience) [inline tests/python/unittest/test_numpy_op.py]: Why 1e0 and 1e-1?

Comment (hzfan) [reply tests/python/unittest/test_numpy_op.py]: When shape is large, the behavior somehow diverges from official NumPy, so after the `np.einsum('abiz,abjz->abij', lhs, rhs)` test case is added (which sums up 512 numbers), I have to relax the tolerance.

Different implementation details can result in the minor difference, like intermediate result type. I require the intermediate var to be of the same type as input vars. But official numpy may has some type promotion here (of which I am not sure). This difference grows as the input size grows.

Also, note that exactly the same result can be produced with int inputs.

Comment (sxjscience) [reply tests/python/unittest/test_numpy_op.py]: I see. Could we compare with some ground-truth values other than the numpy-produced ones? Also, we may set the accumulation type to be float32, which may also improve the performance.

Comment (hzfan) [reply tests/python/unittest/test_numpy_op.py]: Yes, I think it’s a good idea to set the accumulation type. A promotion rule (from input type to accumulation type) may do this, like int32 to int64, float16 to float32, float32 to float64, float64 to float64 etc. I think it can be done in another pr.

As far as I know, MXNet numpy operators have been using official numpy as ground truth. If there are better options, plz let me know.

Comment (sxjscience) [reply tests/python/unittest/test_numpy_op.py]: @hzfan For testing, we always compare with the implementation that is as close to the ground truth as possible. One solution is to benchmark against a higher precision output by numpy, e.g., float64. Setting the `rtol` and `atol` to some large values is not the correct approach. Also, you may set the accumulation type to a higher precision dtype like what we are doing in safe accumulate: https://github.com/apache/incubator-mxnet/pull/14830

Comment (sxjscience) [reply tests/python/unittest/test_numpy_op.py]: @hzfan  For `np.einsum('abiz,abjz->abij', lhs, rhs)`, is it possible to fall back to Batched GEMM? I've communicated with @reminisce offline and einsum is quite important in implementing neural attention. This is critical as users will use it to implement their own model.

Comment (sxjscience) [reply tests/python/unittest/test_numpy_op.py]: @hzfan You may also refer to this tutorial in TF: https://www.tensorflow.org/guide/keras/rnn

Comment (hzfan) [reply tests/python/unittest/test_numpy_op.py]: @sxjscience Falling back to Batched GEMM is possible, but quite a lot work needs to be done. Now both official numpy and our implementation uses `tensordot` as a fallback, but `tensordot` cannot handle batch dot.

Comment (hzfan) [reply tests/python/unittest/test_numpy_op.py]: @sxjscience safe accumulation has been added. Thank you for the suggestion.

**Comment by haojin2** [comment]
@hzfan When you have shrunk some tolerance I would suggest trying multiple trials (like 1000) locally on your side to verify the flakiness a bit before merging, other than that this PR LGTM so far. @sxjscience Any more feedbacks?

**Comment by sxjscience** [comment]
LGTM

**Comment by hzfan** [comment]
@haojin2 I will do it. Now each test takes around 12 minutes on cpu (simply 1 forward takes around 5s, but we have different hybridize, dtype, optimize, forward, backward...), so 1000 tests take around 8.5 days, which is too long.

I will enforce hybridize=True, optimize=False, and use dtypes=`['float16', 'float32', 'float64']`, run 1000 tests. It should finish within a day.

**Comment by haojin2** [comment]
@hzfan Please do that, I think hybridized or not will generally not have effect on the numerical precison. 
Also you could probably use only ~500 trials, a test that succeeds with 99% of chance could only pass 500 trials with less than 1% chance.
